[WIP] Auth

DKhalil opened the pull request:
Create a working auth base

DKhalil committed:
feat(sso): added guardian to deps

DKhalil committed:
chore: updated elixir dependency

DKhalil committed:
feat(api): basic guardian config

DKhalil committed:
feat(api/auth): added ecto type for hashed passwords

DKhalil committed:
feat(sso): added minimum request time to requests to SSO

Fixes #9

igalic commented:
AppVeyor is failing because:

```
  1) test casting custom ecto hashed password cast (SealasSso.EctoHashedPasswordTest)
     test/sealas_sso/types/ectohashedpassword_test.exs:5
     ** (UndefinedFunctionError) function Argon2.Base.hash_password/3 is undefined (module Argon2.Base is not available)
     code: assert {:ok, hash} = EctoHashedPassword.cast("test_password")
     stacktrace:
       (argon2_elixir) Argon2.Base.hash_password("test_password", <<200, 136, 117, 74, 3, 115, 161, 227, 47, 223, 32, 106, 76, 241, 247, 11>>, [])
       (sealas_sso) lib/sealas_sso/types/hashedpassword.ex:7: EctoHashedPassword.cast/1
       test/sealas_sso/types/ectohashedpassword_test.exs:6: (test)
```

this despite argon2 apparently compiling successfully…

igalic commented on apps/sealas_api/mix.exs:
```diff
@@ -9,7 +9,7 @@ defmodule SealasApi.Mixfile do
       deps_path: "../../deps",
       lockfile: "../../mix.lock",
-      elixir: "~> 1.4",
+      elixir: "~> 1.5",
```
huh??

igalic commented on apps/sealas_sso/config/dev.exs:
```diff
@@ -20,6 +20,10 @@ config :sealas_sso, SealasSso.Repo,
 
+config :sealas_sso, SealasSso.Guardian,
+  issuer: "sealas",
+  secret_key: "YQnnTBIJqfn9QA7jp371vqZxKzjxCI9mjmbAWnyjDm013nHqeNp3gE/Zl0k11/b/"
```
how secret is this key really?

igalic commented on apps/sealas_sso/lib/sealas_sso/router.ex:
```diff
@@ -1,13 +1,31 @@
+    Conn.register_before_send(conn, fn conn ->
+      diff = Time.diff(Time.utc_now(), time, :microsecond)
+
+      if diff < @minimum_request_time, do: :timer.sleep round((@minimum_request_time - diff)/1000)
```
shouldn't we log (or fail in) the extraordinarily unlikely event that it takes longer than `@minimum_request_time`

igalic commented on apps/sealas_sso/mix.exs:
```diff
@@ -9,7 +9,7 @@ defmodule SealasSso.Mixfile do
       deps_path: "../../deps",
       lockfile: "../../mix.lock",
-      elixir: "~> 1.4",
+      elixir: "~> 1.5",
```
seriously, how did we not notice this before :D

igalic reviewed (approved):
this looks good so far

DKhalil committed:
chore: hide secret config files everywhere

DKhalil committed:
test(api/ecto): uuid test

DKhalil committed:
feat(sso/auth): added basic auth controller with test

Using encrypted JWTs with elixir's JOSE lib

DKhalil committed:
test(sso/auth): added timer and failed auth tests

DKhalil committed:
feat(sso/views): added 401 and 403 error json

DKhalil committed:
chore: removed warnings

DKhalil committed:
chore: removed warnings

DKhalil committed:
feat(sso/auth): added token auth check in router

DKhalil committed:
test(sso/auth): added tests for requests with tokens

DKhalil committed:
chore: disabled user tests

DKhalil committed:
feat(sso/auth): added yubico dep

DKhalil committed:
feat(sso/auth): added yubico dep

DKhalil committed:
chore: removed unused code

DKhalil committed:
refactor(sso/user): using basemodel now, removing proxy code

DKhalil committed:
doc(sso/router): added some fn docs

DKhalil committed:
chore: lockfile

DKhalil committed:
refactor(sso/user): auth tests using new user schema

DKhalil committed:
feat(sso/tfa): added tfa schema and tests

DKhalil committed:
chore: minor

DKhalil committed:
test(sso/auth): added prelim tests for TFA auth and registration

DKhalil committed:
feat(sso/auth): tfa check at login

DKhalil committed:
feat(sso/tfa): added yubikey validation + tests

DKhalil committed:
feat(config): importing general secret conf

DKhalil committed:
feat(sso/tfa): added skip_server option for testing

DKhalil committed:
feat(sso): added swoosh mailing lib

DKhalil committed:
feat(sso/user): added random password generation fn

DKhalil committed:
feat(sso/tfa): finished tfa implementation with yubikeys

DKhalil committed:
doc(sso/user): added typespecs for schema files

DKhalil committed:
chore: moved auth view to correct dir

DKhalil committed:
doc(sso/auth): added typespecs for auth functions

DKhalil committed:
test(sso/registration): test cases for registration

DKhalil committed:
feat(sso/registration): user registration controller + view

DKhalil committed:
chore: dev db login

DKhalil committed:
chore: disabled user update

DKhalil committed:
feat(sso/auth): login case for unvalidated user

DKhalil committed:
feat(sso/auth): added create_test_changeset

DKhalil committed:
doc(sso/router): added typespecs for router functions

DKhalil committed:
feat(sso/user): added activation_code field to create cast

DKhalil committed:
feat(test): excluding tests with external tag by default

DKhalil committed:
chore: added deps status to readme

igalic committed:
ci: add Makefile to generalize (DRY) the build

on our CIs at least.

igalic committed:
ci: *every* call to mix depends on our secret_config

DKhalil committed:
ci: using nmake on appveyor

DKhalil committed:
ci: fixed config.secret.exs content

igalic committed:
ci: skip travis install step

because it our mix tasks expect the secrets config in place, and travis
doesn't have that.

DKhalil committed:
ci: explicitly install erlang 20.0 before installing elixir

DKhalil committed:
Merge branch 'auth' of https://github.com/DKhalil/sealas into auth
